### ISSUE TITLE
Match error in backend using Java defined annotation with default parameters

### ISSUE THREAD

**Issue by olafurpg** [lampepfl/dotty#2704]
While trying to compile the compiler-benchmark with Dotty I hit on a match error in the backend

```
[error] Error while emitting ColdDotcBenchmark.scala
[error] Block(List(ValDef(warmups$1,TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,scala)),Nothing)],Select(Ident(Fork),$lessinit$greater$default$2)), ValDef(jvm$1,TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,scala)),Nothing)],Select(Ident(Fork),$lessinit$greater$default$3)), ValDef(jvmArgs$1,TypeTree[RefinedType(TypeRef(ThisType(TypeRef(NoPrefix,scala)),Array), scala$Array$$T, TypeAlias(TypeRef(ThisType(TypeRef(NoPrefix,lang)),String), 0))],Apply(Apply(TypeApply(Select(Ident(Array),apply),List(TypeTree[TypeVar(ParamRef(scala$Array$apply$$T) -> TypeRef(ThisType(TypeRef(NoPrefix,lang)),String))])),List(SeqLiteral(List(Literal(Constant(-XX:CICompilerCount=2))),TypeTree[TypeVar(ParamRef(scala$Array$apply$$T) -> TypeRef(ThisType(TypeRef(NoPrefix,lang)),String))]))),List(Apply(TypeApply(Select(Ident(ClassTag),apply),List(TypeTree[TypeVar(ParamRef(scala$Array$apply$$T) -> TypeRef(ThisType(TypeRef(NoPrefix,lang)),String))])),List(Literal(Constant(TypeRef(ThisType(TypeRef(NoPrefix,lang)),String)))))))), ValDef(jvmArgsPrepend$1,TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,scala)),Nothing)],Select(Ident(Fork),$lessinit$greater$default$5))),Apply(Select(New(TypeTree[TypeRef(TermRef(TermRef(TermRef(TermRef(ThisType(TypeRef(NoPrefix,<root>)),org)/withSig(Signature(List(),)),openjdk)/withSig(Signature(List(),)),jmh)/withSig(Signature(List(),)),annotations)/withSig(Signature(List(),)),Fork)]),<init>),List(Literal(Constant(16)), Select(This(Ident(ColdDotcBenchmark)),warmups$1), Select(This(Ident(ColdDotcBenchmark)),jvm$1), Select(This(Ident(ColdDotcBenchmark)),jvmArgs$1), Select(This(Ident(ColdDotcBenchmark)),jvmArgsPrepend$1), Select(Ident(Fork),$lessinit$greater$default$6)))) (of class dotty.tools.dotc.ast.Trees$Block)
scala.MatchError: Block(List(ValDef(warmups$3,TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,scala)),Nothing)],Select(Ident(Fork),$lessinit$greater$default$2)), ValDef(jvm$3,TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,scala)),Nothing)],Select(Ident(Fork),$lessinit$greater$default$3)), ValDef(jvmArgs$3,TypeTree[RefinedType(TypeRef(ThisType(TypeRef(NoPrefix,scala)),Array), scala$Array$$T, TypeAlias(TypeRef(ThisType(TypeRef(NoPrefix,lang)),String), 0))],Apply(Apply(TypeApply(Select(Ident(Array),apply),List(TypeTree[TypeVar(ParamRef(scala$Array$apply$$T) -> TypeRef(ThisType(TypeRef(NoPrefix,lang)),String))])),List(SeqLiteral(List(Literal(Constant(-XX:CICompilerCount=2))),TypeTree[TypeVar(ParamRef(scala$Array$apply$$T) -> TypeRef(ThisType(TypeRef(NoPrefix,lang)),String))]))),List(Apply(TypeApply(Select(Ident(ClassTag),apply),List(TypeTree[TypeVar(ParamRef(scala$Array$apply$$T) -> TypeRef(ThisType(TypeRef(NoPrefix,lang)),String))])),List(Literal(Constant(TypeRef(ThisType(TypeRef(NoPrefix,lang)),String)))))))), ValDef(jvmArgsPrepend$3,TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,scala)),Nothing)],Select(Ident(Fork),$lessinit$greater$default$5))),Apply(Select(New(TypeTree[TypeRef(TermRef(TermRef(TermRef(TermRef(ThisType(TypeRef(NoPrefix,<root>)),org)/withSig(Signature(List(),)),openjdk)/withSig(Signature(List(),)),jmh)/withSig(Signature(List(),)),annotations)/withSig(Signature(List(),)),Fork)]),<init>),List(Literal(Constant(16)), Select(This(Ident(ColdScalacBenchmark)),warmups$3), Select(This(Ident(ColdScalacBenchmark)),jvm$3), Select(This(Ident(ColdScalacBenchmark)),jvmArgs$3), Select(This(Ident(ColdScalacBenchmark)),jvmArgsPrepend$3), Select(Ident(Fork),$lessinit$greater$default$6)))) (of class dotty.tools.dotc.ast.Trees$Block)
    at dotty.tools.backend.jvm.DottyBackendInterface.assocsFromApply(DottyBackendInterface.scala:530)
    at dotty.tools.backend.jvm.DottyBackendInterface$$anon$10.assocs(DottyBackendInterface.scala:521)
    at dotty.tools.backend.jvm.DottyBackendInterface.emitAnnotations$$anonfun$6(DottyBackendInterface.scala:306)
```

With `-Xprint:all`
```scala
[info] result of /Users/ollie/dev/compiler-benchmark/compilation/src/main/scala/dotty/tools/dotc/ColdDotcBenchmark.scala after genBCode:
[info]
[info] package dotty.tools.dotc {
[info]   @scala.annotation.internal.SourceFile(
[info]     "/Users/ollie/dev/compiler-benchmark/compilation/src/main/scala/dotty/tools/dotc/ColdDotcBenchmark.scala"
[info]   ) @{
[info]     <accessor> val warmups$1: Nothing =
[info]       org.openjdk.jmh.annotations.Fork.$lessinit$greater$default$2
[info]     <accessor> val jvm$1: Nothing =
[info]       org.openjdk.jmh.annotations.Fork.$lessinit$greater$default$3
[info]     <accessor> val jvmArgs$1: Array{scala$Array$$T = String} =
[info]       Array.apply[String^](["-XX:CICompilerCount=2" : String^])(
[info]         scala.reflect.ClassTag.apply[String^](classOf[class String])
[info]       )
[info]     <accessor> val jvmArgsPrepend$1: Nothing =
[info]       org.openjdk.jmh.annotations.Fork.$lessinit$greater$default$5
[info]     new org.openjdk.jmh.annotations.Fork(16,
[info]       dotty.tools.dotc.ColdDotcBenchmark.this.warmups$1
[info]     , dotty.tools.dotc.ColdDotcBenchmark.this.jvm$1,
[info]       dotty.tools.dotc.ColdDotcBenchmark.this.jvmArgs$1
[info]     , dotty.tools.dotc.ColdDotcBenchmark.this.jvmArgsPrepend$1,
[info]       org.openjdk.jmh.annotations.Fork.$lessinit$greater$default$6
[info]     )
[info]   } @org.openjdk.jmh.annotations.OutputTimeUnit(MILLISECONDS) @
[info]     org.openjdk.jmh.annotations.BenchmarkMode
[info]   (
[info]     Array.apply[org.openjdk.jmh.annotations.Mode^](
[info]       [SingleShotTime : org.openjdk.jmh.annotations.Mode^]
[info]     )(
[info]       scala.reflect.ClassTag.apply[org.openjdk.jmh.annotations.Mode^](
[info]         classOf[class Mode]
[info]       )
[info]     )
[info]   ) @org.openjdk.jmh.annotations.State(Benchmark) class ColdDotcBenchmark
[info]      extends
[info]    base.BaseDotcBenchmark {
[info]     def <init>(): Unit =
[info]       {
[info]         super()
[info]         ()
[info]       }
[info]     @org.openjdk.jmh.annotations.Benchmark() override def compile(): Unit =
[info]       super.compile()
[info]   }
[info] }
[info]
```

The source file where the error happens ColdDotcBenchmark.scala: https://github.com/lampepfl/compiler-benchmark/blob/82392bc30df11946cb489d45afc2698e3b48bced/compilation/src/main/scala/dotty/tools/dotc/ColdDotcBenchmark.scala#L10-L18

The `@Fork` annotations is defined in Java from the JMH project, https://github.com/jacek-lewandowski/jmh-jl/blob/fc92780b525f6376b2433518b3dc90a5f511feef/jmh-core/src/main/java/org/openjdk/jmh/annotations/Fork.java#L44

I made several attempts to minimize the error with Fork.java and a tiny Foo.scala but didn't manage to reproduce. To reproduce in the environment I hit the error:

- at commit https://github.com/lampepfl/compiler-benchmark/commit/0e282a72f9a7ef4dcafcf7acacd09ecca756b66f
  - git clone https://github.com/olafurpg/compiler-benchmark
  - git checkout bootstrap
- run `sbt compilation/compile`

**Comment by olafurpg** [comment]
I managed to minimize the error into something that scalac compiles OK but dotc reports an error
```scala
// bench.scala
package foo

@Fork(value = 16)
class Bar

// fork.java
package foo;

public @interface Fork {
    int value() default -1;
    int warmups() default -1;
}
```

Running `dotc bench.scala fork.java` reports

```
-- Error: bench.scala:3:0 ------------------------------------------------------
3 |@Fork(value = 16)
  |^^^^^^^^^^^^^^^^
  |missing argument for parameter warmups of constructor Fork: (value: Int, warmups: Int)foo.Fork

one error found
```

I didn't manage to minimize a repro for the match error, but this might be a good start.

**Comment by DarkDimius** [comment]
@olafurpg, thanks for a nice minimization, but is this the same error?
It looks like in your minimization doesn't crash the compiler, but instead reports an error.

**Comment by olafurpg** [comment]
Correct, that minimization does not crash the compiler. I can open another issue if that makes more sense.

**Comment by olafurpg** [comment]
The PR to compile scala/compiler-benchmark with Dotty, https://github.com/scala/compiler-benchmark/pull/31, is currently blocked by this issue (MatchError).

@DarkDimius I opened https://github.com/lampepfl/dotty/issues/2760 for the "missing argument" error.

**Comment by DarkDimius** [comment]
@olafurpg, I'll prepare a fix to unblock you.

**Comment by olafurpg** [comment]
Thanks @DarkDimius To reproduce the MatchError, you can clone this branch https://github.com/olafurpg/compiler-benchmark/tree/dotty2 (commit c30610306e9bc7b5ed0609f0f24ca1054a48d26f) and run

```
> ++0.2.0-bin-20170614-8ddfcaf-NIGHTLY
> compilation/test
...
[error] Error while emitting ScalacBenchmark.scala
[error] Block(List(ValDef(warmups$3,TypeTree[TypeRef(ThisType(TypeRef(NoPrefix,scala)),Nothing)]
```

**Comment by smarter** [comment]
@DarkDimius Is the issue really closed or should it stay open?

**Comment by olafurpg** [comment]
@smarter I tested the fix locally and was no longer able to reproduce the match error, https://github.com/lampepfl/dotty/pull/2762#issuecomment-310436655

**Comment by DarkDimius** [comment]
@smarter I've added logic to backed that tries to figure out what happened to a tree inside an annotation(made it more robust). I didn't investigate who restructured the tree in the first place.
It would be nice to make whoever restructured annotations not do it anymore.

**Comment by smarter** [comment]
Ok, do let's keep this open until we've done that

**Comment by olafurpg** [comment]
Fixed in https://github.com/lampepfl/dotty/pull/2819 ?

**Comment by smarter** [comment]
Yes
